python/ray/data/_internal/execution/resource_manager.py

```diff
@@ -121,3 +136,129 @@ def get_downstream_fraction(self, op: PhysicalOperator) -> float:
         return self._downstream_object_store_memory[op]
+
+    def op_resource_limiter_enabled(self) -> bool:
+        """Return whether OpResourceLimiter is enabled."""
+        return self._op_resource_limiter is not None
+
+    def get_op_limit(self, op: PhysicalOperator) -> ExecutionResources:
+        """Get the limit of resources that the given op can use at the current time.
+
+        This method can only be called when `op_resource_limiter_enabled` returns True.
+        """
+        assert self._op_resource_limiter is not None
+        return self._op_resource_limiter.get_op_limits(op)
+
+
+class OpResourceLimiter(metaclass=ABCMeta):
+    """Interface for limiting resources for each operator.
+
+    This interface allows limit the resources that each operator can use, in each
+    scheduling iteration.
+    """
+
+    def __init__(self, resource_manager: ResourceManager):
+        self._resource_manager = resource_manager
+
+    def on_global_limits_updated(self, global_limits: ExecutionResources):
+        """Callback when the global limits are updated."""
+        pass
+
+    @abstractmethod
+    def update_usages(self) -> ExecutionResources:
+        """Callback to update resource usages."""
+        ...
+
+    @abstractmethod
+    def get_op_limits(self, op: PhysicalOperator) -> ExecutionResources:
+        """Get the limit of resources that the given op can use at the current time."""
+        ...
+
+
+class ReservationOpResourceLimiter(OpResourceLimiter):
+    """An OpResourceLimiter implementation that reserves resources for each operator.
+
+    This class reserves memory and CPU resources for map operators, and consider runtime
+    resource usages to limit the resources that each operator can use.
+
+    It works in the following way:
+    1. Currently we only limit map operators. Non-map operators get unlimited resources.
+    2. For each map operator, we reserve `reservation_ratio * global_resources /
+        num_map_ops` resources. The remaining are shared among all map operators.
+    3. In each scheduling iteration, each map operator will get "remaining of their own
+       reserved resources" + "remaining of shared resources / num_map_ops" resources.
+
+    The `reservation_ratio` is set to 50% by default. Users can tune this value to
+    adjust how aggressive or conservative the resource allocation is. A higher value
+    will make the resource allocation more even, but may lead to underutilization and
+    worse performance. And vice versa.
+    """
+
+    def __init__(self, resource_manager: ResourceManager, reservation_ratio: float):
+        super().__init__(resource_manager)
+        self._reservation_ratio = reservation_ratio
+        assert 0.0 <= self._reservation_ratio <= 1.0
+        # We only limit map operators.
+        self._eligible_ops = [
+            op for op in self._resource_manager._topology if isinstance(op, MapOperator)
+        ]
```

Comment:
Let's add a TODO to handle sub-plan of map operators? i.e. `map().map().repartition().map().map()...`

Reply:
discussed offline, we'll only enable this feature when the dataset only has map/limit/streaming_split. Because memory accounting is not correct for other ops yet.